### PR TITLE
Updating the API guide with a few more pitfalls.

### PR DESCRIPTION
**Pull request by taylanbil**
this is how it looks now: https://github.com/taylanbil/xla/blob/pitfalls/API_GUIDE.md

### CODE REVIEW

File: API_GUIDE.md [hunk 1]
```diff
@@ -45,23 +45,26 @@ for data, target in train_loader:
...
+Note the `xm.mark_step()` line. This is required because of the way XLA tensors work: operations are not executed immediately, but rather added to a graph of pending operations which is only executed when its results are required. Using `xm.mark_step()` acts as an execution barrier which forces the evaluation of the graph accumulated for a single step. Without this barrier, the graph would only be evaluated when evaluating the accuracy of the model, which is only done at the end of an epoch, for this example. Even for small models, the accumulated graph would be too big to evaluate at the end of an entire epoch.
 
 ### MultiCore
 
 There are two ways to drive multiple TPU cores using PyTorch/XLA. One is using the `torch.multiprocessing` module (which internally spawns multiple processes), and the other is using Python threading.
 The multiprocessing method should allow better performance as it gets around the Python GIL serialization, especially with model code which has a heavy Python side processing.
+Note that in the MultiCore setting, `xm.mark_step()` is called inside the data
+iterators, so there is no explicit calls in the examples below.
```
Comment (asuhan) [inline API_GUIDE.md]: "there is no explicit calls" -> "there are no explicit calls"

File: API_GUIDE.md [hunk 2]
```diff
@@ -45,23 +45,26 @@ for data, target in train_loader:
...
+Note the `xm.mark_step()` line. This is required because of the way XLA tensors work: operations are not executed immediately, but rather added to a graph of pending operations which is only executed when its results are required. Using `xm.mark_step()` acts as an execution barrier which forces the evaluation of the graph accumulated for a single step. Without this barrier, the graph would only be evaluated when evaluating the accuracy of the model, which is only done at the end of an epoch, for this example. Even for small models, the accumulated graph would be too big to evaluate at the end of an entire epoch.
```
Comment (asuhan) [inline API_GUIDE.md]: Shouldn't we still mention the `xm.optimizer_step(optimizer)` change from `optimizer.step()`?

File: API_GUIDE.md [hunk 3]
```diff
@@ -143,11 +146,18 @@ PyTorch/XLA behaves semantically like regular PyTorch and XLA tensors, implement
     ```
 
 
-3.  Loops with a different number of iterations between steps are subject to similar observations as tensor shapes. PyTorch/XLA automatically handles them, but they are seen as different execution graphs and require recompilations.
+1.  Loops with a different number of iterations between steps are subject to similar observations as tensor shapes. PyTorch/XLA automatically handles them, but they are seen as different execution graphs and require recompilations. Therefore, to get the best performance, same computations should be run on all XLA devices in all hosts.
```
Comment (asuhan) [inline API_GUIDE.md]: "same computations should be run" -> "same computations should run"

File: API_GUIDE.md (hunk 2 above)
Comment (taylanbil) [inline API_GUIDE.md]: barrier is now in mark_step

File: API_GUIDE.md [hunk 4]
```diff
@@ -143,11 +146,18 @@ PyTorch/XLA behaves semantically like regular PyTorch and XLA tensors, implement
...
+One example of this is, iterators in `torch_xla_py.data_parallel` may drop the
```
Comment (asuhan) [inline API_GUIDE.md]: "For example, iterators in ..." or "One example of this is: iterators in ...".

File: API_GUIDE.md [hunk 5]
```diff
@@ -126,9 +129,9 @@ Check the [full example](https://github.com/pytorch/xla/blob/master/test/test_tr
...
 
 1.  Tensor shapes should be the same between iterations, or a low number of shape variations should be used. PyTorch/XLA automatically recompiles the graph every time new shapes are encountered. This means that, if the shapes don’t stabilize during training, more time will be spent compiling than running the model. Pad tensors to fixed sizes when possible. Direct or indirect uses of `nonzero` introduce dynamic shapes; for example, masked indexing `base[index]` where `index` is a mask tensor.
-2.  Certain operations don’t have native translations to XLA and therefore require transfer to the CPU memory, evaluation on CPU, and transfer of the result back to the XLA device. This is automatically handled by PyTorch/XLA, but doing too many such operations during the training step can lead to significant slowdowns. The `item()` operation is one such example and it is used in [clip_grad_norm_](https://github.com/pytorch/pytorch/blob/de19eeee99a2a282fc441f637b23d8e50c75ecd1/torch/nn/utils/clip_grad.py#L33). Below is an alternative implementation which avoids the need for `item()`:
```
Comment (taylanbil) [inline API_GUIDE.md]: made numberings all 1. 's

File: API_GUIDE.md [hunk 6]
```diff
@@ -143,11 +146,18 @@ PyTorch/XLA behaves semantically like regular PyTorch and XLA tensors, implement
...
+This can dramatically improve performance of the model, up to an N factor, where N is the number of `tensor.item()` calls per step.
 
-4. Sometimes model writers, when knowing that a PyTorch tensor is a scalar, they trigger `tensor.item()` (or equivalent PyTorch APIs which results to the same effects) calls, and they perform operations in Python scalar context, when similar operations can be performed using Pytorch tensor APIs. Following the latter approach will likely result in those operations behind fully fused within an XLA graph, without the need of issuing separate TPU computations.
+`print(torch_xla._XLAC._xla_metrics_report())` can be used to print metrics at the end of each step to collect information regarding the number of compilations and operators that are part of the model but don’t have native XLA implementations. The `XLA_METRICS_FILE=1` environment setting can also be used to export per step metrics to a file.
 
-   This can dramatically improve performance of the model, up to an N factor, where N is the number of `tensor.item()` calls per step.
+One common pitfall is, in this report, any counter that starts with `aten::`
```
Comment (asuhan) [inline API_GUIDE.md]: Is this really a pitfall? The counter merely indicates the pitfall, so I would drop the "One common pitfall is" and start the phrase with "In this report".

File: API_GUIDE.md [hunk 7]
```diff
@@ -143,11 +146,18 @@ PyTorch/XLA behaves semantically like regular PyTorch and XLA tensors, implement
...
+One example of this is, iterators in `torch_xla_py.data_parallel` may drop the
+last few batches in the input iterator, in order to do the same amount of work
+on all XLA devices. In the extreme case where dataset is small, and there are
+too few steps, this may result in a no-op epoch.
 
-    `print(torch_xla._XLAC._xla_metrics_report())` can be used to print metrics at the end of each step to collect information regarding the number of compilations and operators that are part of the model but don’t have native XLA implementations. The `XLA_METRICS_FILE=1` environment setting can also be used to export per step metrics to a file.
+1. Sometimes model writers, when knowing that a PyTorch tensor is a scalar, they trigger `tensor.item()` (or equivalent PyTorch APIs which results to the same effects) calls, and they perform operations in Python scalar context, when similar operations can be performed using Pytorch tensor APIs. Following the latter approach will likely result in those operations behind fully fused within an XLA graph, without the need of issuing separate TPU computations.
```
Comment (asuhan) [inline API_GUIDE.md]: I don't think the repetition of "they" in " they trigger `tensor.item()`" is right. Also, I would rewrite the entire paragraph to be more concise. Something along the lines of: "Even when it's known that a PyTorch tensor is a scalar, avoid using `tensor.item()`. Prefer instead keeping it as a tensor and the use of tensor operations on it, using control flow substitutes such as `torch.where`. Following the latter approach will likely result in those operations behind fully fused within an XLA graph, without the need of issuing separate TPU computations." We can leave this for another pass through the document or defer to a technical writer.

File: API_GUIDE.md (hunk 2 above)
Comment (dlibenzi) [reply API_GUIDE.md]: No, no. This is wrong.
The comment is about optimizer step.
They have to understand that that is needed.
Let's leave this as is for now.
For them, whether the barrier happen in xm.optimizer_step() or later in the data loader, should not matter.
Eventually we can have two separate comments.

File: API_GUIDE.md [hunk 8]
```diff
@@ -45,23 +45,26 @@ for data, target in train_loader:
   loss = loss_fn(output, target)
   loss.backward()
   xm.optimizer_step(optimizer)
```
Comment (dlibenzi) [inline API_GUIDE.md]: It should be:
```Python
xm.optimizer_step(optimizer, barrier=True)
```

Comment (taylanbil) [reply API_GUIDE.md]: I thought about that but then I also thought there may be value in exposing `mark_step()` in the guide. If I do `barrier=True` then we won;t mention mark_step at all. wdyt?

Comment (dlibenzi) [reply API_GUIDE.md]: No please. Let's not ask user to add more lines and understand more offline APIs for the common case.

File: API_GUIDE.md (hunk 1 above)
Comment (dlibenzi) [reply API_GUIDE.md]: Even with multi-processing one should use our data parallel loaders, otherwise device send won't overlap with TPU computations.
I would leave any mentions to mark_step() out for now.

File: API_GUIDE.md (hunk 8 above)
Comment (taylanbil) [reply API_GUIDE.md]: alternative is, include `barrier=True` here and delete that in the multicore cases, and make a note.

File: API_GUIDE.md (hunk 2 above)
Comment (taylanbil) [reply API_GUIDE.md]: The newest version does mention.

File: API_GUIDE.md [hunk 9]
```diff
@@ -126,9 +137,9 @@ Check the [full example](https://github.com/pytorch/xla/blob/master/test/test_tr
...
-2.  Certain operations don’t have native translations to XLA and therefore require transfer to the CPU memory, evaluation on CPU, and transfer of the result back to the XLA device. This is automatically handled by PyTorch/XLA, but doing too many such operations during the training step can lead to significant slowdowns. The `item()` operation is one such example and it is used in [clip_grad_norm_](https://github.com/pytorch/pytorch/blob/de19eeee99a2a282fc441f637b23d8e50c75ecd1/torch/nn/utils/clip_grad.py#L33). Below is an alternative implementation which avoids the need for `item()`:
+1.  Certain operations don’t have native translations to XLA and therefore require transfer to the CPU memory, evaluation on CPU, and transfer of the result back to the XLA device. This is automatically handled by PyTorch/XLA, but doing too many such operations during the training step can lead to significant slowdowns. The `item()` operation is one such example and it is used in [clip_grad_norm_](https://github.com/pytorch/pytorch/blob/de19eeee99a2a282fc441f637b23d8e50c75ecd1/torch/nn/utils/clip_grad.py#L33). Below is an alternative implementation which avoids the need for `item()`:
```
Comment (dlibenzi) [inline API_GUIDE.md]: There is a `1.` already above, no? This should be `2.`?

File: API_GUIDE.md [hunk 10]
```diff
@@ -143,11 +154,23 @@ PyTorch/XLA behaves semantically like regular PyTorch and XLA tensors, implement
     ```
 
 
-3.  Loops with a different number of iterations between steps are subject to similar observations as tensor shapes. PyTorch/XLA automatically handles them, but they are seen as different execution graphs and require recompilations.
+1.  Loops with a different number of iterations between steps are subject to similar observations as tensor shapes. PyTorch/XLA automatically handles them, but they are seen as different execution graphs and require recompilations. Therefore, to get the best performance, same computations should run on all XLA devices in all hosts.
```
Comment (dlibenzi) [inline API_GUIDE.md]: This should be `3.` I think.

Also, there are two different issues.
One is the number of loops which differ across different step.
I would replace the `loops` mentions with one more general along the lines "in order to avoid recompilations, not only shapes must be constant, but also computations" ... and then make the "loop" case be a subcase of that.
That is one issue.

Then you are mixing it with another issue, which is same computation across devices.
There, in training, if the different computation across devices ends up either triggering CRS (Cross Replica Sum) ops with different shapes, or in different sequences, the training would hang.

File: API_GUIDE.md [hunk 11]
```diff
@@ -143,11 +154,23 @@ PyTorch/XLA behaves semantically like regular PyTorch and XLA tensors, implement
...
+One example of this is: iterators in `torch_xla_py.data_parallel` may drop the
```
Comment (dlibenzi) [inline API_GUIDE.md]: This should not be an example of that. Should have it own bullet point, and mention that with very small sample counts, it is better to use small batch sizes.

File: API_GUIDE.md [hunk 12]
```diff
@@ -143,11 +154,23 @@ PyTorch/XLA behaves semantically like regular PyTorch and XLA tensors, implement
...
+   operations on it, using control flow substitutes such as torch.where.
+   Following the latter approach will likely result in those operations behind
+   fully fused within an XLA graph, without the need of issuing separate TPU
+   computations. This can dramatically improve performance of the model, up to
+   an N factor, where N is the number of `tensor.item()` calls per step.
 
-4. Sometimes model writers, when knowing that a PyTorch tensor is a scalar, they trigger `tensor.item()` (or equivalent PyTorch APIs which results to the same effects) calls, and they perform operations in Python scalar context, when similar operations can be performed using Pytorch tensor APIs. Following the latter approach will likely result in those operations behind fully fused within an XLA graph, without the need of issuing separate TPU computations.
+`print(torch_xla._XLAC._xla_metrics_report())` can be used to print metrics at the end of each step to collect information regarding the number of compilations and operators that are part of the model but don’t have native XLA implementations. The `XLA_METRICS_FILE=1` environment setting can also be used to export per step metrics to a file.
```
Comment (dlibenzi) [inline API_GUIDE.md]: It's not `XLA_METRICS_FILE=1` but `XLA_METRICS_FILE=/PATH/TO/FILE`.

File: API_GUIDE.md (hunk 9 above)
Comment (taylanbil) [reply API_GUIDE.md]: markdown handles that, see https://github.com/taylanbil/xla/blob/pitfalls/API_GUIDE.md

This gives us the option to permute the items around in the future.

**Comment by dlibenzi** [comment]
Added @ailzhang to proof validate this as well.

File: API_GUIDE.md [hunk 13]
```diff
@@ -143,11 +154,25 @@ PyTorch/XLA behaves semantically like regular PyTorch and XLA tensors, implement
...
+1. Iterators in `torch_xla_py.data_parallel` may drop the
+last few batches in the input iterator, in order to do the same amount of work
+on all XLA devices. In the extreme case where dataset is small, and there are
+too few steps, this may result in a no-op epoch. Therefore, it is better to use
+small batch sizes in those cases.
 
-    `print(torch_xla._XLAC._xla_metrics_report())` can be used to print metrics at the end of each step to collect information regarding the number of compilations and operators that are part of the model but don’t have native XLA implementations. The `XLA_METRICS_FILE=1` environment setting can also be used to export per step metrics to a file.
+1. Even when it's known that a PyTorch tensor is a scalar, avoid using
+   `tensor.item()`. Prefer instead keeping it as a tensor and the use of tensor
+   operations on it, using control flow substitutes such as torch.where.
```
Comment (asuhan) [inline API_GUIDE.md]: Make "torch.where" inline code (`torch.where`).
